src/test/java/tests/LogicTests.java

```diff
@@ -33,67 +41,115 @@ public LogicTests() throws NoSuchFieldException, IllegalAccessException {
+     * Tests that replaceIssueLabels failed when repoIO failed to update labels
      */
     @Test
-    public void testReplaceIssueLabelsNoChange() throws ExecutionException, InterruptedException {
-        List<String> labels = Arrays.asList("label1", "label2");
-        TurboIssue issue = createIssueWithLabels(labels);
-        mockRepoIOReplaceIssueLabelsResult(labels);
+    public void testReplaceIssueLabelsModelsFailed() throws ExecutionException, InterruptedException {
+        TurboIssue issue = createIssueWithLabels(1, Arrays.asList("label1", "label2"));
+        mockRepoIOReplaceIssueLabelsResult(false);
+        mockMultiModelReplaceIssueLabels(Optional.of(issue), Optional.empty());
 
-        boolean status = logic.replaceIssueLabels(issue, labels).get();
+        assertFalse(logic.replaceIssueLabels(issue, new ArrayList<>()).get());
+    }
 
-        assertTrue(status);
-        assertEquals(labels, issue.getLabels());
+    /**
+     * Tests that {@link MultiModel#replaceIssueLabels(String, int, List)} is first called with the
+     * new labels then revert back to original labels when repoIO failed to update labels
+     */
+    @Test
+    public void testReplaceIssueLabelsRevert() throws ExecutionException, InterruptedException {
+        List<String> originalLabels = Arrays.asList("label1", "label2");
+        List<String> newLabels = Arrays.asList("label3", "label4");
+
+        TurboIssue issue = createIssueWithLabels(1, originalLabels);
+        mockRepoIOReplaceIssueLabelsResult(false);
+        mockMultiModelReplaceIssueLabels(Optional.of(issue), Optional.empty());
+
+        logic.replaceIssueLabels(issue, newLabels).get();
+
+        InOrder inOrder = inOrder(mockedMultiModel);
+        inOrder.verify(mockedMultiModel).replaceIssueLabels(issue.getRepoId(), issue.getId(), newLabels);
+        inOrder.verify(mockedMultiModel).replaceIssueLabels(issue.getRepoId(), issue.getId(), originalLabels);
     }
 
     /**
-     * Tests that the returned status is false when the returned result from RepoIO is inconsistent
-     * with the given labels argument
+     * Tests that no revert is taken place if the issue's labels are modified elsewhere after
+     * {@link Logic#replaceIssueLabels(TurboIssue, List)} is called
      */
     @Test
-    public void testReplaceIssueLabelsInconsistent() throws ExecutionException, InterruptedException {
-        TurboIssue issue = createIssueWithLabels(Arrays.asList("label1", "label2", "label3"));
-        mockRepoIOReplaceIssueLabelsResult(Arrays.asList("label3", "label4"));
+    public void testReplaceIssueLabelsNoRevert() throws ExecutionException, InterruptedException {
+        List<String> originalLabels = Arrays.asList("label1", "label2");
+        List<String> newLabels = Arrays.asList("label3", "label4");
+
+        TurboIssue issue = createIssueWithLabels(1, originalLabels);
+        Thread.sleep(10);
```

Comment:
Maybe the sleep could be encapsulated in a method for creating a 'modified' issue, with a date that can be taken to be after the current time? That way we don't have to repeat it everywhere.